diff-tests/core/core_test.go

```diff
@@ -0,0 +1,458 @@
+
+	// Provider validator status is correct
+	for i := 0; i < len(initState.ValStates.Status); i++ {
+		E := initState.ValStates.Status[i]
+		A := s.validatorStatus(int64(i))
+		if E != A {
+			s.T().Fatal(FAIL_MSG)
+		}
+	}
+
+	// Staking module does not contain undelegations
+	s.providerStakingKeeper().IterateUnbondingDelegations(s.ctx(P),
+		func(index int64, ubd stakingtypes.UnbondingDelegation) bool {
+			s.T().Fatal(FAIL_MSG)
+			return false // Don't stop
+		})
+
+	// Staking module does contain redelegations
+	s.providerStakingKeeper().IterateRedelegations(s.ctx(P),
+		func(index int64, ubd stakingtypes.Redelegation) bool {
+			s.T().Fatal(FAIL_MSG)
+			return false // Don't stop
+		})
+
+	// Staking module does not contain unbonding validators
+	endTime := time.Unix(math.MaxInt64, 0)
+	endHeight := int64(math.MaxInt64)
+	unbondingValIterator := s.providerStakingKeeper().ValidatorQueueIterator(s.ctx(P), endTime, endHeight)
+	defer unbondingValIterator.Close()
+	for ; unbondingValIterator.Valid(); unbondingValIterator.Next() {
+		s.T().Fatal(FAIL_MSG)
+	}
+
+	// Consumer has no slash requests
+	s.Require().Empty(s.consumerKeeper().GetPendingSlashRequests(s.ctx(C)))
+
+	// Consumer has no maturities
+	s.consumerKeeper().IteratePacketMaturityTime(s.ctx(C),
+		func(vscId uint64, timeNs uint64) bool {
+			s.T().Fatal(FAIL_MSG)
+			return false // Don't stop
+		})
+
+	// Consumer power
+	for i := 0; i < len(initState.ValStates.Status); i++ {
+		expectFound := initState.ValStates.Status[i] == stakingtypes.Bonded
+		expectPower := initState.ValStates.Tokens[i]
+		addr := s.validator(int64(i))
+		val, found := s.consumerKeeper().GetCCValidator(s.ctx(C), addr)
+		s.Require().Equal(expectFound, found)
+		if expectFound {
+			if int64(expectPower) != val.Power {
+				s.T().Fatal(FAIL_MSG)
+			}
+		}
+	}
+
+	// The offset time is the last committed time, but the SUT is +1 block ahead
+	// because the currentHeader time is ahead of the last committed. Therefore sub
+	// the difference (duration of 1 block).
+	s.Require().Equal(int64(s.offsetTimeUnix), s.time(P).Add(time.Second*(-6)).Unix())
+	s.Require().Equal(int64(s.offsetTimeUnix), s.time(C).Add(time.Second*(-6)).Unix())
+
+	// The offset height is the last committed height, but the SUT is +1 because
+	// the currentHeader is +1 ahead of the last committed. Therefore sub 1.
+	s.Require().Equal(s.offsetHeight, s.height(P)-1)
+	s.Require().Equal(s.offsetHeight, s.height(C)-1)
+
+	// Network is empty
+	s.Require().Empty(s.simibc.Link.OutboxPackets[P])
+	s.Require().Empty(s.simibc.Link.OutboxPackets[C])
+	s.Require().Empty(s.simibc.Link.OutboxAcks[P])
+	s.Require().Empty(s.simibc.Link.OutboxAcks[C])
+}
+
+// Test a set of traces
+func (s *CoreSuite) TestTraces() {
+	s.traces = Traces{
+		Data: LoadTraces("traces.json"),
+	}
+	for i := range s.traces.Data {
+		s.Run(fmt.Sprintf("Trace num: %d", i), func() {
+			// Setup a new pair of chains for each trace
+			s.SetupTest()
+			s.traces.CurrentTraceIx = i
```

Comment:
This feels really awkward. Can't you just iterate the list and pass the current trace to executeTrace?

Reply:
It might be possible and I was just lazy. I'll check again. The thing is I want to keep the trace in 'global' state for diagnostics